randovania/game_description/game_database_view.py

```diff
@@ -0,0 +1,304 @@
+        """
+        Gets a ResourceInfo of type Damage, using internal name
+        Raises KeyError if it doesn't exist.
+        """
+        raise NotImplementedError
+
+    def get_damage_reduction(self, resource: SimpleResourceInfo, current_resources: ResourceCollection) -> float:
+        """
+        Gets the damage reduction for given resource with the given current resources.
+        """
+        raise NotImplementedError
+
+    def get_template_requirement(self, name: str) -> NamedRequirementTemplate:
+        """
+        Gets a RequirementTemplate, using internal name.
+        Raises KeyError if it doesn't exist.
+        """
+        raise NotImplementedError
+
+    def get_all_resources_of_type(self, resource_type: ResourceType) -> Sequence[ResourceInfo]:
+        """
+        Gets a list of all resources of the given type.
+        """
+        raise NotImplementedError
+
+    def get_pickup_model(self, name: str) -> PickupModel:
+        """
+        Gets a model with the given name.
+        """
+        raise NotImplementedError
+
+
+class ResourceDatabaseViewProxy(ResourceDatabaseView):
+    """
+    A ResourceDatabaseView, implemented by delegating all calls to another ResourceDatabaseView.
+    Intended to be used to overwrite specific functions.
+    """
+
+    def __init__(self, original: ResourceDatabaseView):
+        self._original = original
+
+    @override
+    def get_item(self, short_name: str) -> ItemResourceInfo:
+        return self._original.get_item(short_name)
+
+    @override
+    def get_event(self, short_name: str) -> SimpleResourceInfo:
+        return self._original.get_event(short_name)
+
+    @override
+    def get_trick(self, short_name: str) -> TrickResourceInfo:
+        return self._original.get_trick(short_name)
+
+    @override
+    def get_all_tricks(self) -> Sequence[TrickResourceInfo]:
+        return self._original.get_all_tricks()
+
+    @override
+    def get_damage(self, short_name: str) -> SimpleResourceInfo:
+        return self._original.get_damage(short_name)
+
+    @override
+    def get_damage_reduction(self, resource: SimpleResourceInfo, current_resources: ResourceCollection) -> float:
+        return self._original.get_damage_reduction(resource, current_resources)
+
+    @override
+    def get_template_requirement(self, name: str) -> NamedRequirementTemplate:
+        return self._original.get_template_requirement(name)
+
+    @override
+    def get_all_resources_of_type(self, resource_type: ResourceType) -> Sequence[ResourceInfo]:
+        return self._original.get_all_resources_of_type(resource_type)
+
+
+class GameDatabaseView(ABC):
+    """
+    Provides access to the GameDescription and nested, with support for being filtered for Preset settings.
+
+    These APIs are all expected to be slow and shouldn't be used in any performance sensitive code.
+    """
+
+    @abc.abstractmethod
+    def node_iterator(self) -> Iterable[tuple[Region, Area, Node]]:
```

Comment:
this should probably be an `Iterator`. `RegionList.all_regions_areas_nodes` is an Iterator despite being typed as Iterable, lol